utilities/api/index.ts

```diff
@@ -0,0 +1,60 @@
+import { CONTENTFUL_API_ACCESS_TOKEN, CONTENTFUL_API_URL, HttpRequestMethod } from './constants'
+import type { PageContentProps } from '../../components/Main/PageContent'
+
+type CollectionData = { pageContentCollection: { items: Pathname[] } }
+type ItemData = { pageContentCollection: { items: [PageContentProps] } }
+type Data = CollectionData | ItemData
+export type Pathname = { pathname: string }
+
+async function fetchGraphQl(operationName: string, query: string): Promise<Data> {
+  const res = await fetch(CONTENTFUL_API_URL, {
+    method: HttpRequestMethod.POST,
+    headers: {
+      Authorization: CONTENTFUL_API_ACCESS_TOKEN,
+      'Content-Type': 'application/json',
+    },
+    body: JSON.stringify({
+      operationName,
+      query,
+    }),
+  })
+
+  const { data }: { data: Data } = await res.json()
+
+  return data
+}
+
+export async function getAllPathnames(): Promise<Pathname[] | []> {
+  const operationName = 'GetAllPathnames'
+  const query = `
+    query ${operationName} {
+      pageContentCollection {
+        items {
+          pathname
+        }
+      }
+    }
+  `
+
+  const data: Data = (await fetchGraphQl(operationName, query)) as CollectionData
+
+  return data.pageContentCollection.items
+}
+
+export async function getPageContentByPathname(pathname: string): Promise<PageContentProps | undefined> {
+  const operationName = 'GetPageContentByPathname'
+  const query = `
+    query ${operationName} {
+      pageContentCollection(where: {pathname: "${pathname}"}) {
+        items {
+          content
+          title
+        }
+      }
+    }
+    `
+
+  const data: ItemData = <ItemData>await fetchGraphQl(operationName, query)
+
+  return data.pageContentCollection.items?.[0]
+}
```

Comment:
Here is a cleaner approach to this module.

First, the `api` doesn't belong in `utilities`. It's a top level folder so let's move it up.

Next, let's break the functions out into their own folders each with its own test and use the default export. We can move common types into `types.ts` and constants into `constants.ts`

In `/api/constants.ts`:

```typescript
export const CONTENTFUL_API_ACCESS_TOKEN = 'Bearer qVGC4H5ZmObAQMKyniPJNtNjtGkfmhi_OgVVy9jmtR8'
export const CONTENTFUL_API_URL = 'https://graphql.contentful.com/content/v1/spaces/zqvr29h6rh51/'

export enum HttpRequestMethod {
  GET = 'GET',
  POST = 'POST',
  DELETE = 'DELETE',
  PUT = 'PUT',
}

export const headers = {
  Authorization: CONTENTFUL_API_ACCESS_TOKEN,
  'Content-Type': 'application/json',
}
```

In `/api/types.ts`:

```typescript
export type Path = { pathname: string }
```

In `/api/getPageContentByPathname/index.ts`:

```typescript
import { CONTENTFUL_API_URL, headers, HttpRequestMethod } from '../constants'
import type { PageContentProps } from '../../components/Main/PageContent'

export default async function getPageContentByPathname(pathname: string): Promise<PageContentProps | undefined> {
  const body = JSON.stringify({
    operationName: 'GetPageContentByPathname',
    query: `query GetPageContentByPathname {
      pageContentCollection(where: {pathname: "${pathname}"}) {
        items {
          content
          title
        }
      }
    }`,
  })

  const result = await fetch(CONTENTFUL_API_URL, {
    method: HttpRequestMethod.POST,
    headers,
    body,
  })

  const json = await result.json()
  const [pageContent] = json?.data?.pageContentCollection?.items || []

  return pageContent
}
```

In `/api/getPageContentByPathname/index.test.ts`:

```typescript
import getPageContentByPathname from './'
import { CONTENTFUL_API_URL } from '../constants'
import { rest } from 'msw'
import { setupServer } from 'msw/node'
import 'whatwg-fetch'
import type { PageContentProps } from '../../components/Main/PageContent'

const mockPageContent: PageContentProps = {
  content: 'mock content',
  title: 'mock title',
}

// todo: use mock graphql
const server = setupServer(
  rest.post(CONTENTFUL_API_URL, (req, res, ctx) => {
    const query = (req.body as Record<string, string>).query
    const items: PageContentProps[] = [mockPageContent]

    return res(
      ctx.json({
        data: {
          pageContentCollection: {
            items,
          },
        },
      }),
    )
  }),
)

test('get page content by pathname', async () => {
  server.listen({ onUnhandledRequest: 'error' })

  expect(await getPageContentByPathname('mockPathname')).toMatchObject(mockPageContent)

  server.close()
})
```

In `/api/listPagePaths/index.ts`:

```typescript
import { CONTENTFUL_API_URL, headers, HttpRequestMethod } from '../constants'
import type { Path } from '../types'

export default async function listPagePaths(): Promise<Path[]> {
  const body = JSON.stringify({
    operationName: 'ListPagePaths',
    query: `query ListPagePaths {
      pageContentCollection {
        items {
          pathname
        }
      }
    }`,
  })

  const result = await fetch(CONTENTFUL_API_URL, {
    method: HttpRequestMethod.POST,
    headers,
    body,
  })

  const json = await result.json()

  return json?.data?.pageContentCollection?.items || []
}
```

In `/api/listPagePaths/index.test.ts`:

```typescript
import listPagePaths from './'
import { CONTENTFUL_API_URL } from '../constants'
import { rest } from 'msw'
import { setupServer } from 'msw/node'
import 'whatwg-fetch'
import type { PageContentProps } from '../../components/Main/PageContent'
import type { Path } from '../types'

const mockAllPaths: Path[] = [
  {
    pathname: 'demo/js',
  },
  {
    pathname: 'demo/python',
  },
]

const server = setupServer(
  rest.post(CONTENTFUL_API_URL, (req, res, ctx) => {
    const query = (req.body as Record<string, string>).query
    const items: Path[] = mockAllPaths

    return res(
      ctx.json({
        data: {
          pageContentCollection: {
            items,
          },
        },
      }),
    )
  }),
)

test('list page paths', async () => {
  server.listen({ onUnhandledRequest: 'error' })

  expect(await listPagePaths()).toMatchObject(mockAllPaths)

  server.close()
})
```

In `/pages/_types.ts`:

```typescript
import type { PageContentProps } from '../components/Main/PageContent'

export type SitePageProps = {
  pageContent: PageContentProps
}

export type PageParams = {
  params: {
    slug: string[]
  }
}

export type StaticProps = {
  props: {
    pageContent?: Record<string, string>
  }
}

export type PagePath = {
  params: {
    slug: string[]
  }
}

export type StaticPaths = {
  fallback: boolean
  paths: PagePath[]
}
```

Finally, I had to update `/pages/[[...slug]]/index.ts` thus:

```typescript
import * as React from 'react'
import getPageContentByPathname from '../../api/getPageContentByPathname'
import listPagePaths from '../../api/listPagePaths'
import PageContent from '../../components/Main/PageContent'
import type { Path } from '../../api/types'
import type { PageParams, PagePath, SitePageProps, StaticPaths, StaticProps } from '../_types'

export const getStaticProps = async ({ params: { slug } }: PageParams): Promise<StaticProps> => {
  const pathname: string = slug?.join('/') ?? 'index'
  const pageContent = await getPageContentByPathname(pathname)

  return { props: { pageContent } }
}

export const getStaticPaths = async (): Promise<StaticPaths> => {
  const pagePaths: Path[] = await listPagePaths()

  return {
    fallback: false,
    paths: pagePaths.map<PagePath>(({ pathname }) => ({
      params: {
        slug: pathname === 'index' ? [] : pathname.split('/'),
      },
    })),
  }
}

export default function SitePage({ pageContent }: SitePageProps): JSX.Element {
  return <PageContent content={pageContent.content} title={pageContent.title} />
}
```

I prefer to put the default export at the bottom.

Page works, no errors, tests pass. What did I miss?

I think that this is much cleaner. The names more clearly state what each thing does. Each thing does just one thing. Files are (mostly) shorter and easier to see at a glance. The folder structure better mirrors the module structure. The tests are simpler and we don't need `before` and `after` (I try to avoid them). Etc.

@reubenberghan did I get anything wrong? @natebowang do you see what I'm getting at?

Reply:
All looks pretty good to me, nice and clean. My only note is re the `HttpRequestMethod` enum and potentially moving that into the `types.ts` file and importing accordingly. Unless I'm missing something about that...